xmrstak/backend/cpu/crypto/cryptonight_aesni.h

```diff
@@ -298,53 +326,33 @@ void cryptonight_hash(const void* input, size_t len, void* output, cryptonight_c
 	uint8_t* l0 = ctx0->long_state;
 	uint64_t* h0 = (uint64_t*)ctx0->hash_state;
 
-	uint64_t al0 = h0[0] ^ h0[4];
-	uint64_t ah0 = h0[1] ^ h0[5];
+	__m128i ax0 = _mm_set_epi64x(h0[1] ^ h0[5], h0[0] ^ h0[4]);
 	__m128i bx0 = _mm_set_epi64x(h0[3] ^ h0[7], h0[2] ^ h0[6]);
-
-	uint64_t idx0 = h0[0] ^ h0[4];
+	__m128i cx0 = _mm_set_epi64x(0, 0);
 
 	// Optim - 90% time boundary
-	for(size_t i = 0; i < ITERATIONS; i++)
+	for (size_t i = 0; i < ITERATIONS/2; i++)
 	{
-		__m128i cx;
-		cx = _mm_load_si128((__m128i *)&l0[idx0 & MASK]);
-
-		if(SOFT_AES)
-			cx = soft_aesenc(cx, _mm_set_epi64x(ah0, al0));
-		else
-			cx = _mm_aesenc_si128(cx, _mm_set_epi64x(ah0, al0));
-
-		_mm_store_si128((__m128i *)&l0[idx0 & MASK], _mm_xor_si128(bx0, cx));
-		idx0 = _mm_cvtsi128_si64(cx);
-		bx0 = cx;
-
-		if(PREFETCH)
-			_mm_prefetch((const char*)&l0[idx0 & MASK], _MM_HINT_T0);
-
-		uint64_t hi, lo, cl, ch;
-		cl = ((uint64_t*)&l0[idx0 & MASK])[0];
-		ch = ((uint64_t*)&l0[idx0 & MASK])[1];
-
-		lo = _umul128(idx0, cl, &hi);
-
-		al0 += hi;
```

Comment:
Removal of split a is what I would have an issue with. For single and double hashes it was demonstrably faster in my tests. Can you show the evidence to the contrary?

Reply:
I'm surprised that you observe faster performance with split "a". In my tests I observe virtually no difference in speed. If there is a difference, my version is actually a tad faster.

May I ask what compiler / toolchain are you using?

Mine is g++ 4.2.1 / Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.2.0

I've compared the disassembly of the single hash function both without and with my change, and they look almost identical. Even with the split al0 and ah0, compiler is still smart enough to recognize that it is beneficial to keep the pair {ah0, al0} in one 128-bit xmm register.

For example, prior to the "aesenc" instruction the values {ah0, al0} are always already stored in an xmm and does not require extra bit shuffling.

Also, the code surrounding the 64-bit x 64-bit unsigned multiplication are idential as well. It's always 4 instructions: mulxq, vmovq, vmovq and vpuncklqdq:

```
000000010002e215        c4 e2 cb f6 d6  mulxq   %rsi, %rsi, %rdx
000000010002e21a        c4 e1 f9 6e de  vmovq   %rsi, %xmm3                                                                        
000000010002e21f        c4 e1 f9 6e e2  vmovq   %rdx, %xmm4 
000000010002e224        c5 d9 6c db     vpunpcklqdq     %xmm3, %xmm4, %xmm3    
```

The only difference in the generated code is that, because I'm using merged {ah0, al0}, all memory operations for "ax" are using aligned load/store instructions (vmovdqa) instead of the unaligned (vmovdqu).

Reply:
Comparing single-thread, single-hash mode performance on my device. In both cases I run xmr-stak in a maximized window with no other background app.
```
"cpu_threads_conf" :
[
    { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : false },
],
```
At commit 2bbe9c46 (without my change): 72.8 H/s
At commit 359e7320 (with my change): 72.8 H/s

Reply:
Note that I'm fine either way (if splitting "a" does work better on some compilers). I only felt it was cleaner and easier to read. For Mac OS X it doesn't make much difference.

Reply:
@grzegorzszczecin I would keep the split "a" the way it is for n=1 and n=2 - I actually wrote it with "a" in xmm first time round, however this is slower than having it split.

Reply:
@fireice-uk no problem; I can keep the existing versions of single/double hash functions and only use the CN_STEPx macros for the x3, x4 and x5 versions. Will update the commit later today after some testing.

Reply:
@fireice-uk Uploaded commit 17c5f42 which preserves the split "a" for n=1 and n=2. Thanks for the feedback!